lib/model/model.go

```diff
@@ -850,78 +850,155 @@ func (m *Model) ClusterConfig(deviceID protocol.DeviceID, cm protocol.ClusterCon
 				}
 
-				for _, er := range m.deviceFolders[id] {
-					if er == folder.ID {
-						// We already share the folder with this device, so
-						// nothing to do.
-						continue nextDevice
-					}
+				m.cfg.SetDevice(newDeviceCfg)
+				changed = true
+			}
+
+			for _, er := range m.deviceFolders[id] {
+				if er == folder.ID {
+					// We already share the folder with this device, so
+					// nothing to do.
+					continue nextDevice
 				}
+			}
 
-				// We don't yet share this folder with this device. Add the device
-				// to sharing list of the folder.
+			// We don't yet share this folder with this device. Add the device
+			// to sharing list of the folder.
 
-				l.Infof("Adding device %v to share %q (vouched for by introducer %v)", id, folder.ID, deviceID)
+			l.Infof("Sharing folder %q with %v (vouched for by introducer %v)", folder.ID, id, cfg.DeviceID)
 
-				m.deviceFolders[id] = append(m.deviceFolders[id], folder.ID)
-				m.folderDevices[folder.ID] = append(m.folderDevices[folder.ID], id)
+			m.deviceFolders[id] = append(m.deviceFolders[id], folder.ID)
+			m.folderDevices[folder.ID] = append(m.folderDevices[folder.ID], id)
 
-				folderCfg := m.cfg.Folders()[folder.ID]
-				folderCfg.Devices = append(folderCfg.Devices, config.FolderDeviceConfiguration{
-					DeviceID: id,
-				})
-				m.cfg.SetFolder(folderCfg)
+			folderCfg := m.cfg.Folders()[folder.ID]
+			folderCfg.Devices = append(folderCfg.Devices, config.FolderDeviceConfiguration{
+				DeviceID:     id,
+				IntroducedBy: cfg.DeviceID,
+			})
+			m.cfg.SetFolder(folderCfg)
 
+			changed = true
+		}
+	}
+
+	// If permitted, check if the introducer has unshare devices/folders with
+	// some of the devices/folders that we know were introduced to us by him.
+	if !cfg.SkipIntroductionRemovals {
+
+		// Check if we should unshare some folders, if the introducer has unshared them.
+		for _, folderCfg := range m.cfg.Folders() {
+			folderChanged := false
+		nextIntroducedFolderDevice:
+			for i := 0; i < len(folderCfg.Devices); i++ {
+				if folderCfg.Devices[i].IntroducedBy.Equals(cfg.DeviceID) {
+					for _, device := range foldersDevices[folderCfg.ID] {
+						if bytes.Equal(device.ID, folderCfg.Devices[i].DeviceID[:]) {
+							continue nextIntroducedFolderDevice
+						}
+					}
+					// We could not find that folder shared on the introducer with the device that was introduced to us.
+					// We should follow and unshare aswel.
+					l.Infof("Unsharing folder %q with %v as introducer %v no longer shares the folder with that device", folderCfg.ID, folderCfg.Devices[i].DeviceID, cfg.DeviceID)
+					folderCfg.Devices = append(folderCfg.Devices[:i], folderCfg.Devices[i+1:]...)
+					i--
+					folderChanged = true
+				}
+			}
+
+			// We've modified the folder, hence update it.
+			if folderChanged {
+				m.cfg.SetFolder(folderCfg)
 				changed = true
 			}
 		}
+
+		// Check if we should remove some devices, if the introducer no longer shares any folder with them.
```

Comment:
No I was genuinely concerned. I'm thinking of this scenario:

- We add device `Device` because introducer says so, with folder `Folder` because introducer says so.
- We also share `OtherFolder` with `Device` later, manually
- The `Device` is removed on the introducer.
- We get into this path of the code, and do ???